File: trunk/Yuki/Examples/Communication/CSharp/Src/Server/Servers/TcpSession.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using Firefly;
using BaseSystem;
using Net;

namespace Server
{
    public partial class Tcp<TServerContext>
    {
        /// <summary>
        /// 本类的所有公共成员均是线程安全的。
        /// </summary>
        public class TcpSession
        {
            public TcpServer Server { get; private set; }
            private StreamedAsyncSocket Socket;
            public IPEndPoint RemoteEndPoint { get; private set; }

            private ISessionContext Context;
            private IServerImplementation si;
            private ITcpVirtualTransportServer vts;
            private int NumBadCommands = 0;
            private Boolean IsDisposed = false;

            private Byte[] WriteBuffer;
            private SessionStateMachine<TcpVirtualTransportServerHandleResult, Unit> ssm;

            public TcpSession(TcpServer Server, StreamedAsyncSocket Socket, IPEndPoint RemoteEndPoint)
            {
                this.Server = Server;
                this.Socket = Socket;
                this.RemoteEndPoint = RemoteEndPoint;
                ssm = new SessionStateMachine<TcpVirtualTransportServerHandleResult, Unit>(ex => ex is SocketException, OnCriticalError, OnShutdownRead, OnShutdownWrite, OnWrite, OnExecute, OnStartRawRead, OnExit);
                Socket.TimedOut += ssm.NotifyExit;

                Context = Server.ServerContext.CreateSessionContext();
                Context.Quit += ssm.NotifyExit;
                Context.Authenticated += () => Server.NotifySessionAuthenticated(this);

                if (Server.SerializationProtocolType == SerializationProtocolType.Binary)
                {
                    var p = Server.ServerContext.CreateServerImplementationWithBinaryAdapter(Context);
                    si = p.Key;
                    var a = p.Value;
                    BinaryCountPacketServer.CheckCommandAllowedDelegate cca = CommandName =>
                    {
                        if (Server.CheckCommandAllowed == null) { return true; }
                        return Server.CheckCommandAllowed(Context, CommandName);
                    };
                    var bcps = new BinaryCountPacketServer(a, cca);
                    vts = bcps;
                }
                else if (Server.SerializationProtocolType == SerializationProtocolType.Json)
                {
                    var p = Server.ServerContext.CreateServerImplementationWithJsonAdapter(Context);
                    si = p.Key;
                    var a = p.Value;
                    JsonLinePacketServer.CheckCommandAllowedDelegate cca = CommandName =>
                    {
                        if (Server.CheckCommandAllowed == null) { return true; }
                        return Server.CheckCommandAllowed(Context, CommandName);
                    };
                    vts = new JsonLinePacketServer(a, cca);
                }
                else
                {
                    throw new InvalidOperationException("InvalidSerializationProtocol: " + Server.SerializationProtocolType.ToString());
                }
                vts.ServerEvent += () => ssm.NotifyWrite(new Unit());
                vts.InputByteLengthReport += (CommandName, ByteLength) => Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "InBytes", Name = CommandName, Message = ByteLength.ToInvariantString() });
                vts.OutputByteLengthReport += (CommandName, ByteLength) => Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "OutBytes", Name = CommandName, Message = ByteLength.ToInvariantString() });
            }

            private void OnShutdownRead()
            {
                Socket.Shutdown(SocketShutdown.Receive);
                ssm.NotifyShutdownReadSuccess();
            }
            private void OnShutdownWrite()
            {
                Socket.Shutdown(SocketShutdown.Send);
                ssm.NotifyShutdownWriteSuccess();
            }
            private void OnWrite(Unit w, Action OnSuccess, Action OnFailure)
            {
                var ByteArrays = vts.TakeWriteBuffer();
                if (ByteArrays.Length == 0)
                {
                    OnSuccess();
                    return;
                }
                var TotalLength = ByteArrays.Sum(b => b.Length);
                if ((WriteBuffer == null) || (TotalLength > WriteBuffer.Length))
                {
                    WriteBuffer = new Byte[GetMinNotLessPowerOfTwo(TotalLength)];
                }
                var Offset = 0;
                foreach (var b in ByteArrays)
                {
                    Array.Copy(b, 0, WriteBuffer, Offset, b.Length);
                    Offset += b.Length;
                }
                Socket.SendAsync
                (
                    WriteBuffer,
                    0,
                    TotalLength,
                    OnSuccess,
                    ex =>
                    {
                        if (!IsSocketErrorKnown(ex))
                        {
                            OnCriticalError(ex, new StackTrace(true));
                        }
                        OnFailure();
                    }
                );
            }
            private void OnExecute(TcpVirtualTransportServerHandleResult r, Action OnSuccess, Action OnFailure)
            {
                if (r.OnCommand)
                {
                    var CommandName = r.Command.CommandName;

                    Action a = () =>
                    {
                        var CurrentTime = DateTime.UtcNow;
                        Context.RequestTime = CurrentTime;
                        if (Server.ServerContext.EnableLogPerformance)
                        {
                            var sw = new Stopwatch();
                            sw.Start();
                            r.Command.ExecuteCommand();
                            sw.Stop();
                            Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "Time", Name = CommandName, Message = String.Format("{0}ms", sw.ElapsedMilliseconds) });
                        }
                        else
                        {
                            r.Command.ExecuteCommand();
                        }
                        ssm.NotifyWrite(new Unit());
                    };

                    if (Debugger.IsAttached)
                    {
                        ssm.AddToActionQueue(() =>
                        {
                            a();
                            OnSuccess();
                        });
                    }
                    else
                    {
                        ssm.AddToActionQueue(() =>
                        {
                            try
                            {
                                a();
                            }
                            catch (Exception ex)
                            {
                                RaiseUnknownError(CommandName, ex, new StackTrace(true));
                            }
                            OnSuccess();
                        });
                    }
                }
                else if (r.OnBadCommand)
                {
                    var CommandName = r.BadCommand.CommandName;

                    NumBadCommands += 1;

                    // Maximum allowed bad commands exceeded.
                    if (Server.MaxBadCommands != 0 && NumBadCommands > Server.MaxBadCommands)
                    {
                        RaiseError(CommandName, "Too many bad commands, closing transmission channel.");
                        OnFailure();
                    }
                    else
                    {
                        RaiseError(CommandName, "Not recognized.");
                        OnSuccess();
                    }
                }
                else if (r.OnBadCommandLine)
                {
                    var CommandLine = r.BadCommandLine.CommandLine;

                    NumBadCommands += 1;

                    // Maximum allowed bad commands exceeded.
                    if (Server.MaxBadCommands != 0 && NumBadCommands > Server.MaxBadCommands)
                    {
                        RaiseError("", String.Format(@"""{0}"": Too many bad commands, closing transmission channel.", CommandLine));
                        OnFailure();
                    }
                    else
                    {
                        RaiseError("", String.Format(@"""{0}"":  recognized.", CommandLine));
                        OnSuccess();
                    }
                }
                else
                {
                    throw new InvalidOperationException();
                }
            }
            private void OnStartRawRead()
            {
                Action<int> Completed = Count =>
                {
                    if (Count <= 0)
                    {
                        ssm.NotifyStartRawReadFailure();
                        return;
                    }
                    if (ssm.IsExited()) { return; }
                    var Results = new List<TcpVirtualTransportServerHandleResult>();
                    var c = Count;
                    while (true)
                    {
                        var Result = vts.Handle(c);
                        c = 0;
                        if (Result.OnContinue)
                        {
                            break;
                        }
                        Results.Add(Result);
                    }
                    if (Results.Count == 0)
                    {
                        OnStartRawRead();
                        return;
                    }
                    ssm.NotifyStartRawReadSuccess(Results.ToArray());
                };
                Action<Exception> Faulted = ex =>
                {
                    if (!IsSocketErrorKnown(ex))
                    {
                        OnCriticalError(ex, new StackTrace(true));
                    }
                    ssm.NotifyStartRawReadFailure();
                };
                var Buffer = vts.GetReadBuffer();
                var BufferLength = Buffer.Offset + Buffer.Count;
                Socket.ReceiveAsync(Buffer.Array, BufferLength, Buffer.Array.Length - BufferLength, Completed, Faulted);
            }

            public void Dispose()
            {
                if (IsDisposed) { return; }
                IsDisposed = true;

                IsExitingValue.Update(b => true);
                ssm.NotifyExit();

                Server.SessionMappings.DoAction(Mappings =>
                {
                    if (Mappings.ContainsKey(Context))
                    {
                        Mappings.Remove(Context);
                    }
                });
                Server.ServerContext.TryUnregisterSession(Context);

                si.Dispose();

                IsRunningValue.Update(b => false);

                SpinWait.SpinUntil(() => ssm.IsExited());

                Socket.Dispose();
                Context.Dispose();

                IsExitingValue.Update(b => false);

                if (Server.ServerContext.EnableLogSystem)
                {
                    Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "Sys", Name = "SessionExit", Message = "" });
                }
            }

            private void OnExit()
            {
                IsExitingValue.Update(b =>
                {
                    if (!IsRunningValue.Check(bb => bb)) { return b; }
                    if (!b)
                    {
                        Server.NotifySessionQuit(this);
                    }
                    return true;
                });
            }

            public void Start()
            {
                IsRunningValue.Update
                (
                    b =>
                    {
                        if (b) { throw new InvalidOperationException(); }
                        return true;
                    }
                );

                try
                {
                    Context.RemoteEndPoint = RemoteEndPoint;

                    Server.ServerContext.RegisterSession(Context);
                    Server.SessionMappings.DoAction(Mappings => Mappings.Add(Context, this));

                    if (Server.ServerContext.EnableLogSystem)
                    {
                        Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "Sys", Name = "SessionEnter", Message = "" });
                    }
                    ssm.Start();
                }
                catch (Exception ex)
                {
                    OnCriticalError(ex, new StackTrace(true));
                    ssm.NotifyFailure();
                }
            }

            private LockedVariable<Boolean> IsRunningValue = new LockedVariable<Boolean>(false);
            private LockedVariable<Boolean> IsExitingValue = new LockedVariable<Boolean>(false);
            public Boolean IsRunning
            {
                get
                {
                    return IsRunningValue.Check(b => b);
                }
            }

            private static Boolean IsSocketErrorKnown(Exception ex)
            {
                var sex = ex as SocketException;
                if (sex == null) { return false; }
                var se = sex.SocketErrorCode;
                if (se == SocketError.ConnectionAborted) { return true; }
                if (se == SocketError.ConnectionReset) { return true; }
                if (se == SocketError.Shutdown) { return true; }
                if (se == SocketError.OperationAborted) { return true; }
                if (se == SocketError.Interrupted) { return true; }
                return false;
            }

            private static int GetMinNotLessPowerOfTwo(int v)
            {
                //计算不小于TotalLength的最小2的幂
                if (v < 1) { return 1; }
                var n = 0;
                var z = v - 1;
                while (z != 0)
                {
                    z >>= 1;
                    n += 1;
                }
                var Value = 1 << n;
                if (Value == 0) { throw new InvalidOperationException(); }
                return Value;
            }

            //线程安全
            public void RaiseError(String CommandName, String Message)
            {
                si.RaiseError(CommandName, Message);
            }
            //线程安全
            public void RaiseUnknownError(String CommandName, Exception ex, StackTrace s)
            {
                var Info = ExceptionInfo.GetExceptionInfo(ex, s);
                if (Server.ServerContext.ClientDebug)
                {
                    si.RaiseError(CommandName, Info);
                }
                else
                {
                    si.RaiseError(CommandName, "Internal server error.");
                }
                if (Server.ServerContext.EnableLogUnknownError)
                {
                    Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "Unk", Name = "Exception", Message = Info });
                }
            }

            //线程安全
            private void OnCriticalError(Exception ex, StackTrace s)
            {
                if (Server.ServerContext.EnableLogCriticalError)
                {
                    var Info = ExceptionInfo.GetExceptionInfo(ex, s);
                    Server.ServerContext.RaiseSessionLog(new SessionLogEntry { Token = Context.SessionTokenString, RemoteEndPoint = RemoteEndPoint, Time = DateTime.UtcNow, Type = "Crtcl", Name = "Exception", Message = Info });
                }
            }
        }
    }
}

File: trunk/Yuki/Examples/Communication/CPP/Src/Server/Servers/TcpSession.cpp
#include "TcpSession.h"
#include "TcpServer.h"

#include "BaseSystem/StringUtilities.h"
#include "BaseSystem/Times.h"
#include "Rc4PacketServerTransformer.h"

#include <chrono>
#include <stdexcept>

namespace Server
{
    TcpSession::TcpSession(TcpServer &Server, std::shared_ptr<asio::ip::tcp::socket> Socket, asio::ip::tcp::endpoint RemoteEndPoint, std::function<std::pair<std::shared_ptr<IServerImplementation>, std::shared_ptr<IStreamedVirtualTransportServer>>(std::shared_ptr<ISessionContext>, std::shared_ptr<IBinaryTransformer>)> VirtualTransportServerFactory, std::function<void(std::function<void()>)> QueueUserWorkItem)
        :
        Server(Server),
        Socket(Socket),
        RemoteEndPoint(RemoteEndPoint),
        LastActiveTimeValue(std::chrono::steady_clock::now()),
        NumBadCommands(0),
        IsDisposed(false),
        IsRunningValue(false),
        IsExitingValue(false)
    {
        ssm = std::make_shared<SessionStateMachine<std::shared_ptr<StreamedVirtualTransportServerHandleResult>, Unit>>(
            [](const std::exception &ex) { return dynamic_cast<const asio::system_error *>(&ex) != nullptr; },
            [this](const std::exception &ex) { return OnCriticalError(ex); },
            [this]() { return OnShutdownRead(); },
            [this]() { return OnShutdownWrite(); },
            [this](Unit w, std::function<void()> OnSuccess, std::function<void()> OnFailure) { return OnWrite(w, OnSuccess, OnFailure); },
            [this](std::shared_ptr<StreamedVirtualTransportServerHandleResult> r, std::function<void()> OnSuccess, std::function<void()> OnFailure) { return OnExecute(r, OnSuccess, OnFailure); },
            [this](std::function<void(std::shared_ptr<std::vector<std::shared_ptr<StreamedVirtualTransportServerHandleResult>>>)> OnSuccess, std::function<void()> OnFailure) { return OnStartRawRead(OnSuccess, OnFailure); },
            [this]() { return OnExit(); },
            QueueUserWorkItem
        );

        Context = Server.ServerContext()->CreateSessionContext();
        Context->Quit = [this]() { ssm->NotifyExit(); };
        Context->Authenticated = [this]()
        {
            this->Server.NotifySessionAuthenticated(this->shared_from_this());
        };

        auto rpst = std::make_shared<Rc4PacketServerTransformer>();
        auto Pair = VirtualTransportServerFactory(Context, rpst);
        si = std::get<0>(Pair);
        vts = std::get<1>(Pair);
        Context->SecureConnectionRequired = [=](std::shared_ptr<SecureContext> c)
        {
            rpst->SetSecureContext(c);
        };
        vts->ServerEvent = [this]() { ssm->NotifyWrite(Unit()); };
        vts->InputByteLengthReport = [this](std::wstring CommandName, std::size_t ByteLength)
        {
            auto e = std::make_shared<SessionLogEntry>();
            e->Token = Context->SessionTokenString();
            e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
            e->Time = UtcNow();
            e->Type = L"InBytes";
            e->Name = CommandName;
            e->Message = ToString(ByteLength);
            this->Server.ServerContext()->RaiseSessionLog(e);
        };
        vts->OutputByteLengthReport = [this](std::wstring CommandName, std::size_t ByteLength)
        {
            auto e = std::make_shared<SessionLogEntry>();
            e->Token = Context->SessionTokenString();
            e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
            e->Time = UtcNow();
            e->Type = L"OutBytes";
            e->Name = CommandName;
            e->Message = ToString(ByteLength);
            this->Server.ServerContext()->RaiseSessionLog(e);
        };
    }

    void TcpSession::OnShutdownRead()
    {
        asio::error_code ec;
        Socket->shutdown(asio::socket_base::shutdown_receive, ec);
    }
    void TcpSession::OnShutdownWrite()
    {
        asio::error_code ec;
        Socket->shutdown(asio::socket_base::shutdown_send, ec);
        Socket->close();
    }

    void TcpSession::OnWrite(Unit w, std::function<void()> OnSuccess, std::function<void()> OnFailure)
    {
        auto ByteArrays = vts->TakeWriteBuffer();
        if (ByteArrays.size() == 0)
        {
            OnSuccess();
            return;
        }
        int TotalLength = 0;
        for (auto b : ByteArrays)
        {
            TotalLength += static_cast<int>(b->size());
        }
        if ((WriteBuffer == nullptr) || (TotalLength > static_cast<int>(WriteBuffer->size())))
        {
            WriteBuffer = std::make_shared<std::vector<std::uint8_t>>();
            WriteBuffer->resize(GetMinNotLessPowerOfTwo(TotalLength), 0);
        }
        auto Offset = 0;
        for (auto b : ByteArrays)
        {
            ArrayCopy(*b, 0, *WriteBuffer, Offset, b->size());
            Offset += b->size();
        }
        auto WriteHandler = [=](const asio::error_code &ec, size_t Count)
        {
            if (ec)
            {
                auto ex = asio::system_error(ec);
                if (!IsSocketErrorKnown(ex))
                {
                    OnCriticalError(ex);
                }
                OnFailure();
            }
            else
            {
                OnSuccess();
            }
        };
        asio::async_write(*Socket, asio::buffer(*WriteBuffer, TotalLength), WriteHandler);
    }
    void TcpSession::OnExecute(std::shared_ptr<StreamedVirtualTransportServerHandleResult> r, std::function<void()> OnSuccess, std::function<void()> OnFailure)
    {
        if (r->OnCommand())
        {
            auto CommandName = r->Command->CommandName;

            auto a = [=]()
            {
                auto CurrentTime = UtcNow();
                Context->RequestTime(CurrentTime);
                if (Server.ServerContext()->EnableLogPerformance())
                {
                    auto StartTime = std::chrono::high_resolution_clock::now();
                    auto OnSuccessInner = [=]()
                    {
                        auto EndTime = std::chrono::high_resolution_clock::now();
                        auto Microseconds = std::chrono::duration_cast<std::chrono::microseconds>(EndTime - StartTime).count();
                        auto e = std::make_shared<SessionLogEntry>();
                        e->Token = Context->SessionTokenString();
                        e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
                        e->Time = UtcNow();
                        e->Type = L"Time";
                        e->Name = CommandName;
                        e->Message = ToString(Microseconds) + L"μs";
                        this->Server.ServerContext()->RaiseSessionLog(e);
                        ssm->NotifyWrite(Unit());
                        OnSuccess();
                    };
                    auto OnFailureInner = [=](const std::exception &ex)
                    {
                        RaiseUnknownError(CommandName, ex);
                        OnSuccess();
                    };
                    r->Command->ExecuteCommand(OnSuccessInner, OnFailureInner);
                }
                else
                {
                    auto OnSuccessInner = [=]()
                    {
                        ssm->NotifyWrite(Unit());
                        OnSuccess();
                    };
                    auto OnFailureInner = [=](const std::exception &ex)
                    {
                        RaiseUnknownError(CommandName, ex);
                        OnSuccess();
                    };
                    r->Command->ExecuteCommand(OnSuccessInner, OnFailureInner);
                }
            };

            ssm->AddToActionQueue(a);
        }
        else if (r->OnBadCommand())
        {
            auto CommandName = r->BadCommand->CommandName;

            NumBadCommands += 1;

            // Maximum allowed bad commands exceeded.
            if (Server.MaxBadCommands() != 0 && NumBadCommands > Server.MaxBadCommands())
            {
                RaiseError(CommandName, L"Too many bad commands, closing transmission channel.");
                OnFailure();
            }
            else
            {
                RaiseError(CommandName, L"Not recognized.");
                OnSuccess();
            }
        }
        else if (r->OnBadCommandLine())
        {
            auto CommandLine = r->BadCommandLine->CommandLine;

            NumBadCommands += 1;

            // Maximum allowed bad commands exceeded.
            if (Server.MaxBadCommands() != 0 && NumBadCommands > Server.MaxBadCommands())
            {
                RaiseError(L"", L"\"" + CommandLine + L"\": Too many bad commands, closing transmission channel.");
                OnFailure();
            }
            else
            {
                RaiseError(L"", L"\"" + CommandLine + L"\":  recognized.");
                OnSuccess();
            }
        }
        else
        {
            throw std::logic_error("InvalidOperationException");
        }
    }
    void TcpSession::OnStartRawRead(std::function<void(std::shared_ptr<std::vector<std::shared_ptr<StreamedVirtualTransportServerHandleResult>>>)> OnSuccess, std::function<void()> OnFailure)
    {
        auto Completed = [=](int Count)
        {
            LastActiveTimeValue.Update([](std::chrono::steady_clock::time_point v) { return std::chrono::steady_clock::now(); });
            if (Count <= 0)
            {
                OnFailure();
                return;
            }
            if (ssm->IsExited()) { return; }
            auto Results = std::make_shared<std::vector<std::shared_ptr<StreamedVirtualTransportServerHandleResult>>>();
            auto c = Count;
            while (true)
            {
                std::shared_ptr<StreamedVirtualTransportServerHandleResult> Result;
                try
                {
                    Result = vts->Handle(c);
                }
                catch (std::exception &ex)
                {
                    if (dynamic_cast<std::logic_error *>(&ex) != nullptr)
                    {
                        auto e = std::make_shared<SessionLogEntry>();
                        e->Token = Context->SessionTokenString();
                        e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
                        e->Time = UtcNow();
                        e->Type = L"Known";
                        e->Name = L"Exception";
                        e->Message = s2w(ex.what());
                        Server.ServerContext()->RaiseSessionLog(e);
                    }
                    else if (!IsSocketErrorKnown(ex))
                    {
                        OnCriticalError(ex);
                    }
                    OnFailure();
                    return;
                }
                c = 0;
                if (Result->OnContinue())
                {
                    break;
                }
                Results->push_back(Result);
            }
            if (Results->size() == 0)
            {
                OnStartRawRead(OnSuccess, OnFailure);
                return;
            }
            OnSuccess(Results);
        };
        auto Faulted = [=](const std::exception &ex)
        {
            if (!IsSocketErrorKnown(ex))
            {
                OnCriticalError(ex);
            }
            OnFailure();
        };
        auto Buffer = vts->GetReadBuffer();
        auto BufferLength = vts->GetReadBufferOffset() + vts->GetReadBufferLength();
        auto ReadHandler = [=](const asio::error_code &ec, size_t Count)
        {
            if (ec)
            {
                Faulted(asio::system_error(ec));
            }
            else
            {
                Completed(Count);
            }
        };
        Socket->async_read_some(asio::buffer(Buffer->data() + BufferLength, Buffer->size() - BufferLength), ReadHandler);
    }

    void TcpSession::Stop()
    {
        if (IsDisposed) { return; }
        IsDisposed = true;

        IsExitingValue.Update([](bool b) { return true; });
        ssm->NotifyExit();

        Server.SessionMappings.DoAction([=](std::shared_ptr<std::unordered_map<std::shared_ptr<ISessionContext>, std::shared_ptr<TcpSession>>> Mappings)
        {
            if (Mappings->count(Context) > 0)
            {
                Mappings->erase(Context);
            }
        });
        Server.ServerContext()->TryUnregisterSession(Context);

        si = nullptr;
        vts = nullptr;

        IsRunningValue.Update([](bool b) { return false; });

        while (!ssm->IsExited())
        {
        }

        auto SessionTokenString = Context->SessionTokenString();
        Context = nullptr;

        IsExitingValue.Update([](bool b) { return false; });

        if (Server.ServerContext()->EnableLogSystem())
        {
            auto e = std::make_shared<SessionLogEntry>();
            e->Token = SessionTokenString;
            e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
            e->Time = UtcNow();
            e->Type = L"Sys";
            e->Name = L"SessionExit";
            e->Message = L"";
            Server.ServerContext()->RaiseSessionLog(e);
        }
    }

    TcpSession::~TcpSession()
    {
        Stop();
    }

    void TcpSession::OnExit()
    {
        IsExitingValue.Update([=](bool b)
        {
            if (!IsRunningValue.Check<bool>([=](bool bb) { return bb; })) { return b; }
            if (!b)
            {
                Server.NotifySessionQuit(this->shared_from_this());
            }
            return true;
        });
    }

    void TcpSession::Start()
    {
        IsRunningValue.Update([=](bool b)
        {
            if (b) { throw std::logic_error("InvalidOperationException"); }
            return true;
        });

        try
        {
            Context->RemoteEndPoint(s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port()));

            Server.ServerContext()->RegisterSession(Context);
            Server.SessionMappings.DoAction([=](std::shared_ptr<std::unordered_map<std::shared_ptr<ISessionContext>, std::shared_ptr<TcpSession>>> Mappings) { (*Mappings)[Context] = this->shared_from_this(); });

            if (Server.ServerContext()->EnableLogSystem())
            {
                auto e = std::make_shared<SessionLogEntry>();
                e->Token = Context->SessionTokenString();
                e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
                e->Time = UtcNow();
                e->Type = L"Sys";
                e->Name = L"SessionEnter";
                e->Message = L"";
                Server.ServerContext()->RaiseSessionLog(e);
            }
            ssm->Start();
        }
        catch (std::exception &ex)
        {
            OnCriticalError(ex);
            ssm->NotifyFailure();
        }
    }

    bool TcpSession::IsSocketErrorKnown(const std::exception &ex)
    {
        auto se = dynamic_cast<const asio::system_error *>(&ex);
        if (se == nullptr) { return false; }
        const asio::error_code &ec = se->code();
        if (ec == asio::error::connection_aborted) { return true; }
        if (ec == asio::error::eof) { return true; }
        if (ec == asio::error::operation_aborted) { return true; }
        return false;
    }

    int TcpSession::GetMinNotLessPowerOfTwo(int v)
    {
        //计算不小于TotalLength的最小2的幂
        if (v < 1) { return 1; }
        auto n = 0;
        auto z = v - 1;
        while (z != 0)
        {
            z >>= 1;
            n += 1;
        }
        auto Value = 1 << n;
        if (Value == 0) { throw std::logic_error("InvalidOperationException"); }
        return Value;
    }

    void TcpSession::ArrayCopy(const std::vector<std::uint8_t> &Source, int SourceIndex, std::vector<std::uint8_t> &Destination, int DestinationIndex, int Length)
    {
        if (Length < 0) { throw std::logic_error("InvalidArgument"); }
        if (SourceIndex < 0) { throw std::logic_error("InvalidArgument"); }
        if (DestinationIndex < 0) { throw std::logic_error("InvalidArgument"); }
        if (SourceIndex + Length > static_cast<int>(Source.size())) { throw std::logic_error("InvalidArgument"); }
        if (DestinationIndex + Length > static_cast<int>(Destination.size())) { throw std::logic_error("InvalidArgument"); }
        if (Length == 0) { return; }
        memcpy(&Destination[DestinationIndex], &Source[SourceIndex], Length);
    }

    void TcpSession::RaiseError(std::wstring CommandName, std::wstring Message)
    {
        si->RaiseError(CommandName, Message);
    }
    void TcpSession::RaiseUnknownError(std::wstring CommandName, const std::exception &ex)
    {
        auto Info = s2w(ex.what());
        if (Server.ServerContext()->ClientDebug())
        {
            si->RaiseError(CommandName, Info);
        }
        else
        {
            si->RaiseError(CommandName, L"Internal server error.");
        }
        if (Server.ServerContext()->EnableLogUnknownError())
        {
            auto e = std::make_shared<SessionLogEntry>();
            e->Token = Context->SessionTokenString();
            e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
            e->Time = UtcNow();
            e->Type = L"Unk";
            e->Name = L"Exception";
            e->Message = Info;
            Server.ServerContext()->RaiseSessionLog(e);
        }
    }

    void TcpSession::OnCriticalError(const std::exception &ex)
    {
        if (Server.ServerContext()->EnableLogCriticalError())
        {
            auto Info = s2w(ex.what());
            auto e = std::make_shared<SessionLogEntry>();
            e->Token = Context->SessionTokenString();
            e->RemoteEndPoint = s2w(this->RemoteEndPoint.address().to_string()) + L":" + ToString(this->RemoteEndPoint.port());
            e->Time = UtcNow();
            e->Type = L"Crtcl";
            e->Name = L"Exception";
            e->Message = Info;
            Server.ServerContext()->RaiseSessionLog(e);
        }
    }
}
